File: bitcoin-prototype/python/block.py
from datetime import datetime
import hashlib
import time
import json

class Block:

    def __init__(self, index, nonce, hash, merkle_root,difficulty,  previous_blockhash,transaction=None ):

        self.index = index
        self.time_stamp = datetime.now().timestamp()
        self.data = datetime.now()
        self.nonce = nonce
        self.hash = hash
        self.merkle_root = merkle_root
        self.difficulty = difficulty
        self.previous_blockhash = previous_blockhash
        self.transaction = transaction if transaction else []


    def create_block(self,index, nonce, hash, merkle_root,difficulty,  previous_blockhash,transaction ):

        return Block(index, nonce, hash, merkle_root, difficulty, previous_blockhash, transaction)

    def calculate_hash(self):
        block_string = f"{self.index}{self.time_stamp}{self.data}{self.nonce}{self.hash}{self.merkle_root}{self.difficulty}{self.previous_blockhash}{self.transaction}"
        return hashlib.sha256(block_string.encode()).hexdigest()

    def addTransaction(self, transaction):
        self.transaction.append(transaction)

    def getTransactions(self):
        return self.transaction

    def mine_block(self,difficulty):
        self.difficulty = difficulty
        target = '0' * difficulty

        start_time = time.time()

        count = 1
        while True:
            hash_candidate = self.calculate_hash()
            if hash_candidate[:difficulty] == target:
                self.hash = hash_candidate
                end_time = time.time()
                print(f"Bloco minerado! Hash: {self.hash} \n")
                print(f"Tempo de mineração: {end_time - start_time:.4f} segundos \n")
                print(f"Tempo de mineração: {(end_time - start_time) / 60:.2f} minutos \n")
                break
            else:
                self.nonce +=1
                print(f"{count}",end="\r")
                count +=1
  
    def to_dict(self):
        return {
            "index": self.index,
            "timestamp": self.time_stamp,
            "date": self.data.isoformat(),
            "nonce": self.nonce,
            "difficulty": self.difficulty,
            "hash": self.hash,
            "previous_blockhash": self.previous_blockhash,
            "merkle_root": self.merkle_root,
            "transactions": self.transaction
        }


'''

'''
#block1 = Block(1, 0, '', 'merkle_root_example', 4, '0000', ['transação1', 'transação2'])

'''

'''
#block1.mine_block(6)

'''

'''
#print(json.dumps(block1.to_dict(), indent=4))

File: bitcoin-prototype/java/bitcoin-prototype/src/Block.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Instant;
import java.time.LocalDate;
import java.util.*;

public class Block {

    private int index;
    private long timeStamp;
    private LocalDate data;
    private long nonce;
    private String hash;
    private String markleRoot;
    private int dificulty;
    private String previusBlockHash;
    private List<String> transactions;

    public Block(int index, long nonce, String hash, String markleRoot, int dificulty, String previusBlockHash, List<String> transactions){
        this.index = index;
        this.timeStamp = Instant.now().getEpochSecond();
        this.data = LocalDate.now();
        this.nonce = nonce;
        this.hash = hash;
        this.markleRoot = markleRoot;
        this.dificulty = dificulty;
        this.previusBlockHash = previusBlockHash;
        this.transactions =(transactions != null)? transactions : new ArrayList<>();

    }

    public static Block creatBlock(int index, long nonce, String hash, String markleRoot, int dificulty, String previusBlockHash, List<String> transactions){
        return new Block(index, nonce, hash,markleRoot,dificulty, previusBlockHash, transactions);
    }

    public String calculatehash(){
        String blockString = index + String.valueOf(timeStamp) + data.toString() + nonce + markleRoot + dificulty + previusBlockHash + transactions.toString();

        return sha256(blockString);
    }

    private String sha256(String input){
        try{

            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[]  hashByte = digest.digest(input.getBytes(StandardCharsets.UTF_8));

            StringBuilder hexString = new StringBuilder();

            for (byte b : hashByte){
                hexString.append(String.format("%02x", b) );
            }
            return hexString.toString();

        }catch (Exception e){
            throw new RuntimeException(e);
        }
    }

    public void addTrasaction(String transaction){
        transactions.add(transaction);
    }

    public List<String> getTransactions(){
        return transactions;
    }

    public void mineBlock(int dificulty){
        this.dificulty = dificulty;
        String target = "0".repeat(dificulty);
        long startTime = System.currentTimeMillis();
        int count = 1;

        while (true){
            String hashCandidate = calculatehash();

            if (hashCandidate.startsWith(target)){
                this.hash = hashCandidate;

                long endTime = System.currentTimeMillis();

                System.out.println("Bloco minerado! Hash: " + hash);

                System.out.println("Tempo em minutos: "+(endTime- startTime)/60000.0);

                break;
            }else{
                nonce++;
                System.out.print(count + "\r");
                count++;
            }
        }

    }

    public Map<String, Object> toMap() {
        Map<String, Object> blockMap = new HashMap<>();

        blockMap.put("index", index);
        blockMap.put("timeStamp", timeStamp);
        blockMap.put("data", data);
        blockMap.put("nonce", nonce);
        blockMap.put("hash", hash);
        blockMap.put("dificulty", dificulty);
        blockMap.put("previousBlockHash", previusBlockHash);
        blockMap.put("merkleRoot", markleRoot);
        blockMap.put("transactions", transactions);

        return blockMap;
    }

}
